### PR TITLE
New Process to merge user settings and default settings

### PR DESCRIPTION
**Pull request by jwildfire**

### CODE REVIEW

**Comment by jwildfire** [comment]
Resolves #33 and #35. I also threw in a line for #32.

@samussiah This is ready for your review.

**Comment by jwildfire** [comment]
@samussiah  I updated the branch to include the new filter functionality.

File: src/chart/init.js [hunk 1]
```diff
@@ -11,65 +11,9 @@ export function init(data) {
...
-            return e.key;
-        }));
-
-  //Set 'Total' column color to #777.
-    if (settings.defaults.totalCol === 'Show')
-        this.colorScale.range()[settings.groups.length] = '#777';
-
   //Initialize adverse event eplorer.
+    this.util.setDefaults(this)
+    console.log(this)
```
Comment (samussiah) [inline src/chart/init.js]: stray

File: src/chart/util/setDefaults.js [hunk 2]
```diff
@@ -0,0 +1,117 @@
...
+        chart.config.defaults.totalCol = false;
+        chart.config.groups = [{'key':'All'}];
+    }
+
+    ////////////////////////////////////////////////////
+    // Include all group levels if none are specified //
+    ////////////////////////////////////////////////////
+    var groups = d3.set(chart.raw_data.map(d => d[chart.config.variables.group])).values();
+    var groupsObject = groups.map(d => {return {'key': d}; });
+
```
Comment (samussiah) [inline src/chart/util/setDefaults.js]: I like this.

File: src/chart/util/setDefaults.js [hunk 3]
```diff
@@ -0,0 +1,117 @@
...
+    
+    function errorNote(msg) {
+        chart.wrap.append('div').attr('class', 'alert').text("Fatal Error: "+msg);
+    };
+
+    /////////////////////////////
+    // Fill defaults as needed //
+    /////////////////////////////
+    //variables
+    chart.config.variables = chart.config.variables || {}
```
Comment (samussiah) [inline src/chart/util/setDefaults.js]: Smart.

File: src/chart/util/setDefaults.js [hunk 4]
```diff
@@ -0,0 +1,117 @@
...
+    function errorNote(msg) {
+        chart.wrap.append('div').attr('class', 'alert').text("Fatal Error: "+msg);
+    };
+
+    /////////////////////////////
+    // Fill defaults as needed //
+    /////////////////////////////
+    //variables
+    chart.config.variables = chart.config.variables || {}    
+    var variables = ["id","major","minor","group","details"]
```
Comment (samussiah) [inline src/chart/util/setDefaults.js]: It would be a good idea to include filter variables here as well.

File: src/chart/util/setDefaults.js [hunk 5]
```diff
@@ -0,0 +1,117 @@
...
+            .sort((a,b) => a.key < b.key ? -1 : a.key > b.key ? 1 : 0);
+
+    //////////////////////////////////////////////////////////////
+    //Check that variables specified in settings exist in data. //
+    //////////////////////////////////////////////////////////////
+    for (var x in chart.config.variables) {
+        var varList = d3.keys(chart.raw_data[0]).concat('data_all');
+
+        if (varList.indexOf(chart.config.variables[x]) === -1) {
+            if (chart.config.variables[x] instanceof Array) {
```
Comment (samussiah) [inline src/chart/util/setDefaults.js]: Add filters to chart.config.variables above, mapped to keep only the `value_col` property.

File: src/chart/util/setDefaults.js [hunk 6]
```diff
@@ -0,0 +1,117 @@
...
+            if (groups.indexOf(d.key) === -1) {
+                errorNote('Error in settings object.');
+                throw new Error('\'' + e.key + '\' in the Groups setting is not found in the dataset.');
+            }
+        });
+    
+    /////////////////////////////////////////////////////////////////////////////////
+    //Check that group values defined in settings are actually present in dataset. //
+    /////////////////////////////////////////////////////////////////////////////////
+            if (chart.config.groups.length > 6) {
```
Comment (samussiah) [inline src/chart/util/setDefaults.js]: Curious, how'd you arrive 6?

File: src/chart/defaultSettings.js [hunk 7]
```diff
@@ -18,18 +18,14 @@ const defaultSettings =
...
         []
     ,'defaults':
         {'maxPrevalence': 0
-        ,'totalCol': 'Show'
-        ,'diffCol': 'Show'
-        ,'prefTerms': 'Hide'}
+        ,'totalCol': true
+        ,'diffCol': true
+        ,'prefTerms': false}
     ,'plotSettings':
```
Comment (samussiah) [inline src/chart/defaultSettings.js]: Interesting.  Applies only to the rate plots?

File: src/chart/util/setDefaults.js (hunk 6 above)
Comment (jwildfire) [inline src/chart/util/setDefaults.js]: No good reason, just seemed about right. Would probably be better to make this a setting actually  `config.defaults.maxGroups` - will make that change.

File: src/chart/util/setDefaults.js (hunk 4 above)
Comment (jwildfire) [inline src/chart/util/setDefaults.js]: I'm on board. will move `settings.filters` to `settings.variables.filters`

**Comment by jwildfire** [comment]
@samussiah Changes are in. Look good?

**Comment by samussiah** [comment]
Damn. Skippy.

**Pull request merged**